compiler/noirc_evaluator/src/acir/arrays.rs

```diff
@@ -814,24 +814,42 @@ impl Context<'_> {
         if !array_typ.contains_slice_element() {
             array_typ.flattened_size() as usize
         } else {
-            let mut size = 0;
             match &dfg[array] {
-                Value::NumericConstant { .. } => {
-                    size += 1;
-                }
-                Value::Instruction { .. } => {
+                Value::NumericConstant { .. } => 1,
+                Value::Instruction { .. } | Value::Param { .. } => {
                     let array_acir_value = self.convert_value(array, dfg);
-                    size += flattened_value_size(&array_acir_value);
+                    flattened_value_size(&array_acir_value)
                 }
-                Value::Param { .. } => {
+                _ => {
+                    unreachable!("ICE: Unexpected SSA value when computing the slice size");
+                }
+            }
+        }
+    }
+
+    /// Check if the array or slice has 0 length.
+    ///
+    /// This is different from `flattened_size` in that a non-zero length
+    /// array containing zero length arrays has zero size, but we can still
+    /// access its elements.
+    fn has_zero_length(&mut self, array: ValueId, dfg: &DataFlowGraph) -> bool {
+        if let Type::Array(_, size) = dfg.type_of_value(array) {
+            size == 0
+        } else {
+            match &dfg[array] {
+                Value::NumericConstant { constant, .. } => constant.is_zero(),
```

Comment:
This is unreachable, no? An array can't be a numeric constant.

Reply:
Removed this match so it's `unreachable!` now.